src/components/form/adapters/base-adapter.ts

```diff
@@ -0,0 +1,211 @@
+import React from 'react';
+import { isNil, isNumber, isString, isBoolean } from 'lodash-es';
+
+const toDashCase = (myStr) =>
+    myStr.replace(/([A-Z])/g, (g) => `-${g[0].toLowerCase()}`);
+
+// Checks if value is a primitive
+const isPrimitive = (value) =>
+    isNil(value) || isNumber(value) || isString(value) || isBoolean(value);
+
+/**
+ * Renders a webcomponent by name
+ */
+export class LimeElementsAdapter extends React.Component<any, any> {
+    component = null;
+
+    static defaultProps = {
+        // The events to bind the the web component
+        events: {},
+        // The props to pass to the web component
+        elementProps: {},
+    };
+
+    constructor(
+        public props: {
+            name: string;
+            events?: any;
+            elementProps?: any;
+        }
+    ) {
+        super(props);
+
+        this.state = {
+            prevEvents: {},
+        };
+
+        this.component = React.createRef();
+
+        this.setComponentProperty = this.setComponentProperty.bind(this);
+    }
+
+    // Sets a property on the web component
+    setComponentProperty([key, value]: [string, any]) {
+        const element = this.component.current;
+        element[key] = value;
+    }
+
+    /**
+     * Get the handler for an event.
+     *
+     * The handler stop propagation otherwise. Should be used instead of just binding
+     * the handler passed in the `events` prop due to this reason
+     *
+     * @param {string} eventName the name of the event to get the handler for
+     *
+     * @returns {any} the handler to bind to the component for the event name
+     */
+    getHandler(eventName: string) {
+        const { events } = this.props;
+        return (event) => {
+            event.stopPropagation();
+            events[eventName](event);
+        };
+    }
+
+    /**
+     * Set events on the web component
+     *
+     * @param {object} events Event names to handlers
+     *
+     * @returns {void}
+     */
+    setEvents(events: object) {
+        const element = this.component.current;
+        for (const event of Object.keys(events)) {
+            element.addEventListener(event, this.getHandler(event));
+        }
+    }
+
+    /**
+     * Remove events on the web component
+     *
+     * @param {object} events Event names to handlers
+     *
+     * @returns {void}
+     */
+    removeEvents(events: object) {
+        const element = this.component.current;
+        for (const event of Object.keys(events)) {
+            element.removeEventListener(event, events[event]);
+        }
+    }
+
+    /**
+     * Detect if event handlers have changed. If they
+     * have changed, store them in the state, and remove and
+     * re-add the changed event handlers
+     *
+     * @returns {void}
+     */
+    updateEvents() {
```

Comment:
Do you think we could simplify this method a bit? It has a lot of responsibilities right now. It filters out old events, new events, remove old listeners and sets new listeners. I get what the method does, but I have to read it a few times to fully understand it